src/software/proto/proto_logger/proto_log_reader.h

```diff
@@ -0,0 +1,86 @@
+#pragma once
+#include <experimental/filesystem>
+
+#include "software/proto/repeated_any_msg.pb.h"
+#include "software/util/typename/typename.h"
+
+
+class ProtoLogReader
+{
+   public:
+    /**
+     * Constructs a `ProtoLogReader` that will read numerically-named files containing
+     * a delimited protobuf message with type `RepeatedAnyMsg`, which itself contains
+     * multiple `Any`s, each of which can contain a Protobuf message of any type.
+     *
+     * @param _replay_dir The directory that we want to read replay proto messages from
+     */
+    explicit ProtoLogReader(const std::string& _replay_dir);
+
+    /**
+     * Returns the next `Any` message from the current `RepeatedAnyMsg` message, unpacked
+     * into the desired type `Msg`. Throws std::invalid_argument if unsuccessful.
+     *
+     * @return the next recorded SensorProto if available, nullopt otherwise.
+     */
+    template <typename Msg>
+    std::optional<Msg> getNextMsg();
+
+   private:
+    /**
+     * Returns true if another recorded message is available, and false otherwise.
+     *
+     * @return true if there is another message available
+     */
+    bool hasNextMsg() const;
+
+    /**
+     * Increments the chunk index currently being played from, and reads it from the file.
+     */
+    void nextChunk();
+
+    /**
+     * Reads a delimited RepeatedAnyMsg protobuf message from a file located at the given
+     * path.
+     *
+     * @param file_path the path of the file containing a delimited protobuf file.
+     * @return a RepeatedAnyMsg object that has been read from the given path.
+     */
+    static RepeatedAnyMsg readDelimitedRepeatedAnyMsgFile(
+        const std::experimental::filesystem::path& file_path);
+
+    size_t max_chunk_idx;
+    size_t cur_chunk_idx;
+    int cur_msg_idx;
+    RepeatedAnyMsg cur_chunk;
+    std::experimental::filesystem::path replay_dir;
+};
+
+template <typename Msg>
+std::optional<Msg> ProtoLogReader::getNextMsg()
```

Comment:
is the idea to use one `ProtoLogReader` to read multiple message types? If not, then it would more consistent to make it a template class

Reply:
I'm not a big fan of making things templates unless necessary, for the sake of compile times. In this class, this is the only method that requires the type of the message to be converted into, so only this method was templated.